### PR TITLE
Fix smoke tests

### LINKED ISSUE #2036: Smoke tests failing on SRE workspace
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **managing** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** `workspace_software @ 0fd9565`

## :package: Packages

<details>
<summary>List of packages</summary>

```none
acme==2.6.0
annotated-types==0.6.0
appdirs==1.4.4
Arpeggio==2.0.2
attrs==23.2.0
azure-common==1.1.28
azure-core==1.30.1
azure-identity==1.16.1
azure-keyvault-certificates==4.8.0
azure-keyvault-keys==4.9.0
azure-keyvault-secrets==4.8.0
azure-mgmt-automation==1.0.0
azure-mgmt-compute==30.6.0
azure-mgmt-containerinstance==10.1.0
azure-mgmt-core==1.4.0
azure-mgmt-dns==8.1.0
azure-mgmt-keyvault==10.3.0
azure-mgmt-msi==7.0.0
azure-mgmt-network==25.3.0
azure-mgmt-rdbms==10.1.0
azure-mgmt-resource==23.0.1
azure-mgmt-storage==21.1.0
azure-storage-blob==12.19.1
azure-storage-file-datalake==12.14.0
azure-storage-file-share==12.15.0
build==1.2.1
certifi==2024.2.2
cffi==1.16.0
charset-normalizer==3.3.2
chevron==0.14.0
click==8.1.7
cryptography==42.0.6
-e git+ssh://git@github.com/jemrobinson/data-safe-haven.git@0fd9565306842d98e1c4c6153d626473e41f6d30#egg=data_safe_haven
dill==0.3.8
dnspython==2.4.2
fqdn==1.5.1
freezegun==1.5.1
grpcio==1.60.1
idna==3.7
iniconfig==2.0.0
isodate==0.6.1
josepy==1.14.0
markdown-it-py==3.0.0
mdurl==0.1.2
msal==1.28.0
msal-extensions==1.1.0
msrest==0.7.1
oauthlib==3.2.2
packaging==24.0
parver==0.5
pip-tools==7.4.1
pluggy==1.5.0
portalocker==2.8.2
protobuf==4.25.3
psycopg==3.1.18
pulumi==3.115.0
pulumi_azure_native==2.50.1
pulumi_azuread==5.52.0
pulumi_random==4.16.1
pulumi_tls==5.0.4
pycparser==2.22
pydantic==2.7.1
pydantic_core==2.18.2
Pygments==2.18.0
PyJWT==2.8.0
pyOpenSSL==24.1.0
pyproject_hooks==1.1.0
pyRFC3339==1.1
pytest==8.2.0
pytest-asyncio==0.23.6
pytest-mock==3.14.0
python-dateutil==2.9.0.post0
pytz==2023.4
PyYAML==6.0.1
requests==2.31.0
requests-oauthlib==2.0.0
rich==13.7.1
semver==2.13.0
setuptools==69.5.1
shellingham==1.5.4
simple-acme-dns==3.0.0
six==1.16.0
typer==0.12.3
typing_extensions==4.11.0
urllib3==2.2.1
validators==0.22.0
websocket-client==1.8.0
wheel==0.43.0
```

</details>

## :no_entry_sign: Describe the problem

```bash
root@shm-green-sre-bottle-vm-workspace-01:/home/dshadmin/tests# bats run_all_tests.bats                                                          ✓ Mounted drives (/data) 
 ✓ Mounted drives (/home) 
 ✓ Mounted drives (/output) 
 ✓ Mounted drives (/shared) 
 ✗ Python package repository
   (from function `initialise_python_environment' in file run_all_tests.bats, line 9,
    in test file run_all_tests.bats, line 66)
     `initialise_python_environment' failed
   The virtual environment was not created successfully because ensurepip is not
   available.  On Debian/Ubuntu systems, you need to install the python3-venv
   package using the following command.
   
       apt install python3.10-venv
   
   You may need to use sudo with that command.  After installing the python3-venv
   package, recreate your virtual environment.
   
   Failing command: /root/.local/bats-python-environment/bin/python
   
 ✗ R package repository
   (in test file run_all_tests.bats, line 73)
     `[ "$status" -eq 0 ]' failed with status 127
 ✗ Python functionality
   (from function `initialise_python_environment' in file run_all_tests.bats, line 9,
    in test file run_all_tests.bats, line 80)
     `initialise_python_environment' failed
   The virtual environment was not created successfully because ensurepip is not
   available.  On Debian/Ubuntu systems, you need to install the python3-venv
   package using the following command.
   
       apt install python3.10-venv
   
   You may need to use sudo with that command.  After installing the python3-venv
   package, recreate your virtual environment.
   
   Failing command: /root/.local/bats-python-environment/bin/python
   
 ✓ R functionality 
 - MS SQL database (Python) (skipped: No database credentials available)
 - MS SQL database (R) (skipped: No database credentials available)
 - Postgres database (Python) (skipped: No database credentials available)
 - Postgres database (R) (skipped: No database credentials available)

12 tests, 3 failures, 4 skipped
```

## :steam_locomotive: Workarounds or solutions

1. Only test one Python version
2. Provide DB credentials if appropriate
3. Put the tests in a common location (rather than in a home directory which
  - might fail to mount 
  - might get written to `/home` on the local disk which is then overwritten by the remote mount

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [ ] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [ ] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [ ] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :closed_umbrella: Related issues

Closes #2036 

### :microscope: Tests

Run on a fresh deployment

```
root@shm-daimyo-sre-hojo-vm-workspace-01:/usr/local/smoke_tests# ./run_all_tests.bats 
 ✓ Mounted drives (/data) 
 ✓ Mounted drives (/home) 
 ✓ Mounted drives (/output) 
 ✓ Mounted drives (/shared) 
 ✓ Python package repository 
 ✓ R package repository 
 ✓ Python functionality 
 ✓ R functionality 
 - MS SQL database (Python) (skipped: No database credentials available)
 - MS SQL database (R) (skipped: No database credentials available)
 - Postgres database (Python) (skipped: No database credentials available)
 - Postgres database (R) (skipped: No database credentials available)
```

### CODE REVIEW

**Comment by JimMadge** [comment]
@jemrobinson I'm not certain what to do for the DB credentials.

There is some stuff here I don't like that we should look to change in the future,

- Tests need to be run from a particular directory (?)
- Testing block list packages is disabled
- DB tests are disabled

But I would suggest we don't change that here because,

- These aren't regressions
- We may want to replaces this with an Ansible playbook which would also integrate into our CI

**Comment by jemrobinson** [comment]
Since we're using shared DB credentials (rather than per-user ones as in the past) and we need a way to make these available to (appropriate) users in the workspace anyway - could we write them to a file that's only readable by the "Privileged Users" group?

**Comment by JimMadge** [comment]
Yes, I'm sure we can do that. Where are those credentials? Azure secrets?

Might be easiest (for now) to write them using cloud init

**Comment by jemrobinson** [comment]
Yes, defined in the `data` component [here](https://github.com/alan-turing-institute/data-safe-haven/blob/a21263dc7be1fb3bf889d85ca5025ffb1fba3128/data_safe_haven/infrastructure/programs/sre/data.py#L331-L350). Available either as a `pulumi.Random` string or as a `KeyVault Secret`.

**Comment by JimMadge** [comment]
Looks like there is a configuration problem somewhere,

```
psycopg.OperationalError: connection failed: connection to server at "10.0.1.237", port 5432 failed: FATAL:  password authentication failed for user "databaseadmin@shm-daimyo-sre-hojo-db-server-postgresql"
connection to server at "10.0.1.237", port 5432 failed: FATAL:  no pg_hba.conf entry for host "fd40:9c3a:1b:d897:6f1b:1000:a00:205", user "databaseadmin@shm-daimyo-sre-hojo-db-server-postgresql", database "postgres", no encryption
```

Should check,

- credential is correct
- credential written to workspace correctly

**Comment by jemrobinson** [comment]
`pg_hba.conf entry` makes me think that the **database server** isn't configured to accept connections from this VM.

**Comment by JimMadge** [comment]
@jemrobinson would that be something wrong with my configuration.

I think I deployed the databases after the workspaces, so perhaps the provisioning step didn't run correctly?

**Comment by jemrobinson** [comment]
> Would that be something wrong with my configuration?

No, I think it's a bit of the infrastructure we haven't really tested. Here are the networking rule for the server that I think you're talking about.

Screenshot 2024-07-26 at 10 32 11

I think these need to have public access turned off and a firewall rule that allows access from the SRE (either the whole SRE or just the workspaces subnet) created. We might want a rule that allows access from admin IP addresses too, but I think it's OK to say that moving data from `ingress` into the database is delegated to the users.

Should all be possible to do in Pulumi but I haven't tested it.

**Comment by JimMadge** [comment]
Possibly related [stack overflow question](https://stackoverflow.com/questions/62301317/azure-database-for-postgresql-server-no-pg-hba-conf-entry-for-host).

@jemrobinson Do you think we should review/merge this as is or fix the DB here too?

**Comment by JimMadge** [comment]
Disabling `require_secure_transport` fixed the `pg_hba.conf` message.
Password authentication is still failing.

**Comment by JimMadge** [comment]
Can connect using `psql -h postgresql.hojo.daimyo.develop.turingsafehaven.ac.uk -p 5432 -U databaseadmin postgres`

**Comment by JimMadge** [comment]
Looks like the credential hasn't written correctly 🤔.

**Comment by JimMadge** [comment]
```
# ./run_all_tests.bats
 ✓ Mounted drives (/data) 
 ✓ Mounted drives (/home) 
 ✓ Mounted drives (/output) 
 ✓ Mounted drives (/shared) 
 ✓ Python package repository 
 ✓ R package repository 
 ✓ Python functionality 
 ✓ R functionality 
 ✓ MS SQL database (Python) 
 ✓ MS SQL database (R) 
 ✓ Postgres database (Python) 
 ✓ Postgres database (R) 
```

File: data_safe_haven/infrastructure/programs/declarative_sre.py [hunk 1]
```diff
@@ -350,6 +350,7 @@ def __call__(self) -> None:
                 apt_proxy_server_hostname=apt_proxy_server.hostname,
                 data_collection_rule_id=monitoring.data_collection_rule_vms.id,
                 data_collection_endpoint_id=monitoring.data_collection_endpoint.id,
+                database_service_admin_password=data.password_database_service_admin,
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/programs/declarative_sre.py:353]: Has something weird happened with the indentation here?

File: data_safe_haven/infrastructure/programs/sre/data.py [hunk 2]
```diff
@@ -254,7 +254,7 @@ def __init__(
             tags=child_tags,
         )
 
-        # Secret: database service admin password
+        # Secret: dns server admin password
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/programs/sre/data.py]:
```suggestion
        # Secret: DNS server admin password
```

**Review by jemrobinson — APPROVED** [review]
LGTM. Do we not need `username@hostname` usernames to log in any more?

File: data_safe_haven/infrastructure/programs/declarative_sre.py (hunk 1 above)
Comment (JimMadge) [reply data_safe_haven/infrastructure/programs/declarative_sre.py:353]: Not that I can see, am I missing something?

**Comment by JimMadge** [comment]
> LGTM. Do we not need `username@hostname` usernames to log in any more?

Not for the Postgres database, you need just the username. MSSQL is the same as before.
Not certain if that is a difference in Postgres server or client (or configuration).

**Pull request merged**